File: mobile/lib/components/scanner_sim.dart
import 'package:flutter/material.dart';

class ScannerSim extends StatefulWidget {
  final List<Map<String, dynamic>> targets;
  final VoidCallback onClose;
  final void Function(Map<String, dynamic> object, String type) onScanSuccess;

  const ScannerSim({
    super.key,
    required this.targets,
    required this.onClose,
    required this.onScanSuccess,
  });

  @override
  State<ScannerSim> createState() => _ScannerSimState();
}

class _ScannerSimState extends State<ScannerSim> {
  Map<String, dynamic>? selectedTarget;

  @override
  void initState() {
    super.initState();
    selectedTarget = widget.targets.isNotEmpty ? widget.targets[0] : null;
  }

  void handleSimulateScan() {
    if (selectedTarget == null) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Aucune cible à scanner disponible à proximité')),
      );
      return;
    }
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Scan Réussi'),
        content: Text(
            'Équipement détecté : ${selectedTarget!['type'].toUpperCase()} ${selectedTarget!['object']['code'] ?? selectedTarget!['object']['fid']}'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: const Text('Annuler'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              widget.onScanSuccess(selectedTarget!['object'], selectedTarget!['type']);
            },
            child: const Text('Ouvrir la fiche'),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF1a1a1a),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                const Text(
                  '📷 Scanner QR Code / Code Équipement',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 15,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                TextButton(
                  onPressed: widget.onClose,
                  child: const Text(
                    'Fermer',
                    style: TextStyle(color: Color(0xFFe74c3c)),
                  ),
                ),
              ],
            ),
            const SizedBox(height: 20),
            Expanded(
              flex: 2,
              child: Container(
                decoration: BoxDecoration(
                  color: Colors.black,
                  borderRadius: BorderRadius.circular(8),
                ),
                child: Center(
                  child: Stack(
                    alignment: Alignment.center,
                    children: [
                      Container(
                        width: 180,
                        height: 180,
                        decoration: BoxDecoration(
                          border: Border.all(color: Colors.white30),
                        ),
                        child: Stack(
                          children: [
                            Positioned(
                              top: 0,
                              left: 0,
                              child: Container(
                                width: 20,
                                height: 20,
                                decoration: const BoxDecoration(
                                  border: Border(
                                    top: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                    left: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                  ),
                                ),
                              ),
                            ),
                            Positioned(
                              top: 0,
                              right: 0,
                              child: Container(
                                width: 20,
                                height: 20,
                                decoration: const BoxDecoration(
                                  border: Border(
                                    top: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                    right: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                  ),
                                ),
                              ),
                            ),
                            Positioned(
                              bottom: 0,
                              left: 0,
                              child: Container(
                                width: 20,
                                height: 20,
                                decoration: const BoxDecoration(
                                  border: Border(
                                    bottom: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                    left: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                  ),
                                ),
                              ),
                            ),
                            Positioned(
                              bottom: 0,
                              right: 0,
                              child: Container(
                                width: 20,
                                height: 20,
                                decoration: const BoxDecoration(
                                  border: Border(
                                    bottom: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                    right: BorderSide(width: 4, color: Color(0xFF2ecc71)),
                                  ),
                                ),
                              ),
                            ),
                            const Positioned(
                              top: 0,
                              bottom: 0,
                              child: SizedBox(
                                width: double.infinity,
                                child: Divider(color: Color(0xFFe74c3c), thickness: 2),
                              ),
                            ),
                          ],
                        ),
                      ),
                      const Text(
                        'Placer le code au centre du viseur',
                        style: TextStyle(color: Colors.white38, fontSize: 12),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            const SizedBox(height: 20),
            Expanded(
              flex: 1,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'Équipements détectés par la caméra :',
                    style: TextStyle(color: Colors.white, fontSize: 13),
                  ),
                  const SizedBox(height: 10),
                  if (widget.targets.isNotEmpty)
                    SizedBox(
                      height: 110,
                      child: ListView.builder(
                        itemCount: widget.targets.length,
                        itemBuilder: (context, index) {
                          final tgt = widget.targets[index];
                          return GestureDetector(
                            onTap: () => setState(() => selectedTarget = tgt),
                            child: Container(
                              margin: const EdgeInsets.only(bottom: 6),
                              padding: const EdgeInsets.all(10),
                              decoration: BoxDecoration(
                                color: selectedTarget == tgt
                                    ? const Color(0xFF2ecc71)
                                    : const Color(0xFF2a2a2a),
                                borderRadius: BorderRadius.circular(6),
                              ),
                              child: Text(
                                '${tgt['type'] == 'regard' ? '🕳️ Regard' : '🚰 Conduite'} : ${tgt['object']['code'] ?? tgt['object']['fid']}',
                                style: TextStyle(
                                  color: selectedTarget == tgt
                                      ? Colors.black
                                      : Colors.white70,
                                ),
                              ),
                            ),
                          );
                        },
                      ),
                    )
                  else
                    const Text(
                      'Aucun équipement visible à proximité.',
                      style: TextStyle(
                        color: Colors.white38,
                        fontStyle: FontStyle.italic,
                      ),
                    ),
                  const SizedBox(height: 15),
                  SizedBox(
                    width: double.infinity,
                    child: ElevatedButton(
                      onPressed: selectedTarget != null ? handleSimulateScan : null,
                      style: ElevatedButton.styleFrom(
                        backgroundColor: selectedTarget != null
                            ? const Color(0xFF2ecc71)
                            : Colors.grey[600],
                      ),
                      child: const Text(
                        '⚡ Simuler la détection QR',
                        style: TextStyle(fontWeight: FontWeight.bold),
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

File: mobile/components/ScannerSim.js
/**
 * Nom Auteur : Dr Abdelhakim Kellouche
 * Nom de l'application : Plateforme de Pilotage du Réseau d'Assainissement
 * Numéro version : 1.0.0
 * Date de création : 30-05-2026
 * Date de modification : 30-05-2026
 * 
 * Objectif du module :
 * Simulateur de scanner de QR code / code-barres pour la démonstration mobile.
 * Offre un viseur dynamique et permet de sélectionner un objet du réseau 
 * à scanner pour ouvrir instantanément sa fiche terrain.
 */

import React, { useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Alert,
  ScrollView,
} from 'react-native';

export default function ScannerSim({ targets, onScanSuccess, onClose }) {
  const [selectedTarget, setSelectedTarget] = useState(targets && targets.length > 0 ? targets[0] : null);

  const handleSimulateScan = () => {
    if (!selectedTarget) {
      Alert.alert('Info', 'Aucune cible à scanner disponible à proximité.');
      return;
    }

    Alert.alert(
      'Scan Réussi',
      `Équipement détecté : ${selectedTarget.type.toUpperCase()} ${selectedTarget.object.code || selectedTarget.object.fid}`,
      [
        {
          text: 'Ouvrir la fiche',
          onPress: () => onScanSuccess(selectedTarget.object, selectedTarget.type),
        }
      ]
    );
  };

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>📷 Scanner QR Code / Code Équipement</Text>
        <TouchableOpacity onPress={onClose}>
          <Text style={styles.closeBtn}>Fermer</Text>
        </TouchableOpacity>
      </View>

      {/* Viseur de caméra simulé */}
      <View style={styles.cameraPreview}>
        <View style={styles.reticle}>
          <View style={styles.cornerTopLeft} />
          <View style={styles.cornerTopRight} />
          <View style={styles.cornerBottomLeft} />
          <View style={styles.cornerBottomRight} />
          <View style={styles.laser} />
        </View>
        <Text style={styles.scanText}>Placer le code au centre du viseur</Text>
      </View>

      {/* Zone de contrôle */}
      <View style={styles.controlPanel}>
        <Text style={styles.label}>Équipements détectés par la caméra :</Text>
        
        {targets && targets.length > 0 ? (
          <ScrollView style={styles.targetList}>
            {targets.map((tgt, idx) => (
              <TouchableOpacity
                key={idx}
                style={[
                  styles.targetItem,
                  selectedTarget === tgt && styles.targetItemActive
                ]}
                onPress={() => setSelectedTarget(tgt)}
              >
                <Text style={[styles.targetText, selectedTarget === tgt && styles.targetTextActive]}>
                  {tgt.type === 'regard' ? '🕳️ Regard' : '🚰 Conduite'} : {tgt.object.code || tgt.object.fid}
                </Text>
              </TouchableOpacity>
            ))}
          </ScrollView>
        ) : (
          <Text style={styles.emptyText}>Aucun équipement visible à proximité.</Text>
        )}

        <TouchableOpacity
          style={[styles.scanButton, !selectedTarget && styles.scanButtonDisabled]}
          onPress={handleSimulateScan}
          disabled={!selectedTarget}
        >
          <Text style={styles.scanButtonText}>⚡ Simuler la détection QR</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#1a1a1a', padding: 16 },
  header: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', marginBottom: 20 },
  title: { color: '#fff', fontSize: 15, fontWeight: 'bold' },
  closeBtn: { color: '#e74c3c', fontWeight: 'bold', fontSize: 14 },
  cameraPreview: { flex: 1.2, backgroundColor: '#000', borderRadius: 8, justifyContent: 'center', alignItems: 'center', position: 'relative', overflow: 'hidden' },
  reticle: { width: 180, height: 180, borderWidth: 1, borderColor: 'rgba(255,255,255,0.3)', position: 'relative' },
  cornerTopLeft: { position: 'absolute', top: -2, left: -2, width: 20, height: 20, borderTopWidth: 4, borderLeftWidth: 4, borderColor: '#2ecc71' },
  cornerTopRight: { position: 'absolute', top: -2, right: -2, width: 20, height: 20, borderTopWidth: 4, borderRightWidth: 4, borderColor: '#2ecc71' },
  cornerBottomLeft: { position: 'absolute', bottom: -2, left: -2, width: 20, height: 20, borderBottomWidth: 4, borderLeftWidth: 4, borderColor: '#2ecc71' },
  cornerBottomRight: { position: 'absolute', bottom: -2, right: -2, width: 20, height: 20, borderBottomWidth: 4, borderRightWidth: 4, borderColor: '#2ecc71' },
  laser: { position: 'absolute', top: '50%', left: 0, right: 0, height: 2, backgroundColor: '#e74c3c', opacity: 0.8 },
  scanText: { color: '#aaa', fontSize: 12, marginTop: 20, fontWeight: 'bold' },
  controlPanel: { flex: 1, marginTop: 20 },
  label: { color: '#fff', fontSize: 13, fontWeight: 'bold', marginBottom: 10 },
  targetList: { maxHeight: 110, marginBottom: 15 },
  targetItem: { backgroundColor: '#2a2a2a', padding: 10, borderRadius: 6, marginBottom: 6 },
  targetItemActive: { backgroundColor: '#2ecc71' },
  targetText: { color: '#ccc', fontSize: 13 },
  targetTextActive: { color: '#000', fontWeight: 'bold' },
  emptyText: { color: '#888', fontStyle: 'italic', fontSize: 12, marginBottom: 20 },
  scanButton: { backgroundColor: '#2ecc71', padding: 14, borderRadius: 8, alignItems: 'center', marginTop: 5 },
  scanButtonDisabled: { backgroundColor: '#444' },
  scanButtonText: { color: '#fff', fontWeight: 'bold', fontSize: 14 }
});
